### ISSUE TITLE
I would like to inquire about streaming write and the use of multiple fonts.

### ISSUE THREAD

**Issue by walter-chequer** [phax/ph-pdf-layout#42]
1. When generating a PDF using ph-pdf-layout, if there is a large amount of data to write, does calling pageLayout.renderTo() write everything to the file at once while holding all the data in memory?
Can I save a PDF, then load it again and use saveIncremental?

3. Is there a way to apply a font from a list of registered fonts in order if a character is not supported by the specified font in PLText? In other words, the method would involve registering multiple fonts and applying them sequentially in case of a failure.

thank you :)

**Comment by phax** [comment]
@walter-chequer  thanks for the questions.
ad 1. Currently `pageLayout.renderTo` really does it all at once. However, there is some nice caching built-in to make sure that fonts are loaded as little as possible. So far I heard no complaints about memory usage.
As I was interested myself, I added some more fonts to `PLTextTest.testCustomFontMultiple` and ran it through the profile.
In that test, using 23 different external fonts, the used memory size of the `PLPageLayout` object is ~170MB in YourKit profiles:
![grafik](https://github.com/user-attachments/assets/29242cbd-7ace-4013-b838-cab106e294dc)
This is more than "nothing", but I think acceptable if you consider the size of the raw TTF/OTF fonts.

An incremental approach has not been considered yet. Currently you can only create the documents from scratch. However, there might be interesting use cases, if you want to extend an existing one. Is that something you consider from the perspective of memory saving, or do you indeed want to add one or more pages to an existing PDF document?

Ad 2. Nope, this feature is currently not supported either. And that would be tricky to implement, because the `LoadedFont` has a cache fot the encoded code points based on a specific font (for efficiency reasons this is absolutely necessary).
Not even an optional `Supplier`-like getter providing and alternative `EncodedCodePoint` would really help, as the encoding for the page stream is focuses on one font atm.
So to sum up: currently the effort would be too high

hth, Philip

**Comment by walter-chequer** [comment]
@phax 
The data to be written to the PDF is very large. Therefore, I inquired whether it is possible to generate the PDF while saving data to disk in the middle of the process, as I cannot hold all the data in memory.

In my tests, generating a PDF of about 4MB required 168MB of memory. I found that creating multiple 4MB PDFs and then merging them later using mergeDocuments uses less memory than generating a single PDF all at once.

```kotlin
val pdfMerger = PDFMergerUtility()
pdfMerger.destinationFileName = "merged.pdf"

pdfMerger.addSource(File(source1))
pdfMerger.addSource(File(source2))
pdfMerger.mergeDocuments(IOUtils.createTempFileOnlyStreamCache(), CompressParameters.NO_COMPRESSION)
```

Regarding the font issue, it seems that using a unified font like the one found at https://github.com/satbyy/go-noto-universal could resolve the problem.
thank you :)

**Comment by phax** [comment]
Super. See https://github.com/phax/ph-fonts/ on how to embedd arbitrary fonts - you don't need to wait for me to do anything :) I only have the Noto subset for Simplified Chinese and Traditional Chinese.

Indeed, the merging afterwards saves memory, but the resulting PDF file might be slightly bigger than it would be otherwise. But I guess that is not of the essence in your case.

Happy PDF creation :)
